cpp/include/cuspatial/experimental/detail/linestring_distance.cuh

```diff
@@ -0,0 +1,284 @@
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @tparam Cart2dItB Iterator to 2d cartesian coordinates. Must meet requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @tparam OffsetIterator Iterator to linestring offsets. Must meet requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @tparam OutputIterator Iterator to output distances. Must meet requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible and mutable.
+ *
+ * @param[in] linestring1_offsets_begin Iterator to the begin of the range of linestring offsets
+ * in pair 1.
+ * @param[in] linestring1_offsets_end Iterator to the end of the range of linestring offsets
+ * in pair 1.
+ * @param[in] linestring1_points_xs_begin Iterator to the begin of the range of x coordinates of
+ * points in pair 1.
+ * @param[in] linestring1_points_xs_end Iterator to the end of the range of x coordiantes of points
+ * in pair 1.
+ * @param[in] linestring2_offsets_begin Iterator to the begin of the range of linestring offsets
+ * in pair 2.
+ * @param[in] linestring2_points_xs_begin Iterator to the begin of the range of x coordinates of
+ * points in pair 2.
+ * @param[in] linestring2_points_xs_end Iterator to the end of the range of x coordiantes of points
+ * in pair 2.
+ * @param[out] distances Iterator to the output range of shortest distances between pairs.
+ *
+ * [LinkLRAI]: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+ * "LegacyRandomAccessIterator"
+ */
+template <typename Cart2dItA, typename Cart2dItB, typename OffsetIterator, typename OutputIterator>
+void __global__ pairwise_linestring_distance_kernel(OffsetIterator linestring1_offsets_begin,
+                                                    OffsetIterator linestring1_offsets_end,
+                                                    Cart2dItA linestring1_points_begin,
+                                                    Cart2dItA linestring1_points_end,
+                                                    OffsetIterator linestring2_offsets_begin,
+                                                    Cart2dItB linestring2_points_begin,
+                                                    Cart2dItB linestring2_points_end,
+                                                    OutputIterator distances)
+{
+  using T = typename std::iterator_traits<Cart2dItA>::value_type::value_type;
+
+  auto const p1_idx = threadIdx.x + blockIdx.x * blockDim.x;
+  size_t const num_linestrings =
+    thrust::distance(linestring1_offsets_begin, linestring1_offsets_end);
+
+  size_t const linestring1_num_points =
+    thrust::distance(linestring1_points_begin, linestring1_points_end);
+  size_t const linestring2_num_points =
+    thrust::distance(linestring2_points_begin, linestring2_points_end);
+
+  if (p1_idx >= linestring1_num_points) { return; }
```

Comment:
This exit can technically happen a few lines earlier, but it probably doesn't matter.

Reply:
I think it's cleaner to keep the definitions of `linestring1_num_points` and `linestring2_num_points` together. The compiler can handle this optimization.